src/main/java/de/tum/in/www1/artemis/service/programming/JavaTemplateUpgradeService.java

```diff
@@ -0,0 +1,217 @@
+
+            // Validate that template and repository have the same number of pom.xml files, otherwise no upgrade will take place
+            // TODO: Improve matching of repository and template poms, support sequential test runs
+            if (templatePoms.length == 1 && repositoryPoms.size() == 1) {
+                Model updatedRepoModel = upgradeProjectObjectModel(templatePoms[0], repositoryPoms.get(0));
+                writeProjectObjectModel(updatedRepoModel, repositoryPoms.get(0));
+                programmingExerciseService.commitAndPushRepository(repository, "Template upgraded by Artemis", userService.getUser());
+            }
+        }
+        catch (IOException | GitAPIException | InterruptedException | XmlPullParserException exception) {
+            log.error("Updating of template files of repository " + repositoryType.name() + " for exercise " + exercise.getId() + " failed with error" + exception.getMessage());
+            // Rollback local changes in case of errors
+            gitService.deleteLocalRepository(exercise.getRepositoryURL(repositoryType));
+        }
+    }
+
+    private void writeProjectObjectModel(Model repositoryModel, File repositoryPom) throws IOException {
+        try (OutputStream outputStream = new FileOutputStream(repositoryPom)) {
+            var pomWriter = new MavenXpp3Writer();
+            pomWriter.write(outputStream, repositoryModel);
+        }
+    }
+
+    private Model upgradeProjectObjectModel(Resource templatePom, File repositoryPom) throws IOException, XmlPullParserException {
+        try (InputStream templateInput = templatePom.getInputStream(); InputStream repoInput = new FileInputStream(repositoryPom)) {
+
+            var pomReader = new MavenXpp3Reader();
+            Model templateModel = pomReader.read(templateInput);
+            Model repoModel = pomReader.read(repoInput);
+
+            // Update all dependencies found in the repository pom with version found in the template
+            updateDependencies(repoModel, templateModel);
+
+            // Update Maven Compiler Plugin with JDK version, Maven Surefire Plugin and Maven Failsafe Plugin by replacing them with template plugins
+            upgradePlugin(repoModel, templateModel, "org.apache.maven.plugins", "maven-compiler-plugin");
+            upgradePlugin(repoModel, templateModel, "org.apache.maven.plugins", "maven-surefire-plugin");
+            upgradePlugin(repoModel, templateModel, "org.apache.maven.plugins", "maven-failsafe-plugin");
+
+            // Replace JUnit4 with AJTS
+            removeDependency(repoModel, "junit", "junit");
+            addDependency(repoModel, templateModel, "de.tum.in.ase", "artemis-java-test-sandbox");
+
+            // Remove legacy dependencies which are no longer needed or were moved to AJTS
+            removeDependency(repoModel, "org.json", "json");
```

Comment:
This should remove the dependencies also from the TEST repos right?

Reply:
org.json:json should be always deleted for all repository types.